model/riscv_sys_control.sail

```diff
@@ -337,6 +342,71 @@ val rvfi_trap : unit -> unit
 function rvfi_trap () = ()
 $endif
 
+/* Disable trigger firing in trap handler */
+function TriggerFireMatch(cur_priv : Privilege) -> bool = {
+  /* Second solution of reenterancy problem */
+  if (sys_reent_opt1()) then {
+    match cur_priv {
+      Machine => bits_to_bool(tcontrol[mte]),
+      _       => true
+    }
+  }
+  /* First solution of reenterancy problem */
+  else {
+    match cur_privilege {
+      Machine    => bits_to_bool(mstatus[MIE]),
+      Supervisor => if (medeleg[Breakpoint] == 0b1) then { bits_to_bool(mstatus[SIE]) } else { true },
+      _          => true
+    }
+  }
+}
+
+/* Instruction Count Trigger Match */
+function icount_trigger_match(cur_priv : Privilege) -> unit = {
+  /* Traverse through all the triggers */
+  foreach (i from 0 to (sizeof(N_TRIGGERS) - 1)) {
+    let trigger_type : bits(4) = tdata1_x.tdata1[i][(xlen - 1) .. (xlen - 4)];
+    /* If Instruction Count trigger is found */
+    if (trigger_type == triggerType_to_bits(TRIG_ICOUNT)) then {
+      var InstrCount : InstructionCount  = Mk_InstructionCount(tdata1_x.tdata1[i][26 .. 0]);
+      /* Check that trigger matches */
+      if ((((InstrCount[m] == 0b1) & (cur_priv == Machine))     | /* Trigger is enabled in M-mode */
+           ((InstrCount[s] == 0b1) & (cur_priv == Supervisor))  | /* Trigger is enabled in S-mode */
+           ((InstrCount[u] == 0b1) & (cur_priv == User)) )      & /* Trigger is enabled in U-mode */
+          (InstrCount[count] != zero_extend(14, 0b0))           &
+          (TriggerFireMatch(cur_priv))) then { /* Trigger Match valid? check re-enterancy */
+        /* Decrement the counter */
+        if (InstrCount[count] == zero_extend(14, 0b1)) then {
+          InstrCount[pending] = 0b1;
+        };
+        InstrCount[count]           = InstrCount[count] - 1;
```

Comment:
This looks like count will keep decrementing even if pending is set. The spec says:
> When count is 0 it stays at 0 until explicitly written.

Reply:
The condition is there, to prevent count from decrementing if the value is 0.

https://github.com/YazanHussnain-10x/sail-riscv/blob/92c63e1d1978d8e97c567b28206e390064218f29/model/riscv_sys_control.sail#L328

Reply:
The condition is there to prevent count form decrementing when the value is 0.

> line 328: (InstrCount[count] != zero_extend(14, 0b0))